### PR TITLE
fix: partition and metadata column insertion order in ExtendedColumnProjector

### PR DESCRIPTION
**Pull request by phillipleblanc**
The project() method was inserting partition columns first, then metadata columns, both using their original schema indices. This caused incorrect column positions when partition and metadata columns were interleaved in the schema, since inserting at lower indices shifts subsequent positions.

Fix: Merge partition and metadata column inserts into a single list sorted by schema index, then insert in ascending order.

Also fix file_stream.rs test that expected the buggy behavior, and add regression tests covering interleaved schemas and edge cases.

### CODE REVIEW

File: datafusion/datasource/src/file_stream.rs [hunk 1]
```diff
@@ -1223,16 +1223,17 @@ mod tests {
             .unwrap();
         assert_eq!(id_array.value(0), 1);
 
+        // Check year column (partition value)
         let year_idx = *field_indices.get("year").unwrap();
         let year_col = projected_batch.column(year_idx);
         let year_array = year_col.as_any().downcast_ref::<StringArray>().unwrap();
-        assert_eq!(year_array.value(0), "a");
+        assert_eq!(year_array.value(0), "2023");
```
Comment (sgrebnov) [inline datafusion/datasource/src/file_stream.rs:1230]: 👍

**Review by sgrebnov — APPROVED** [review]

**Review by lukekim — APPROVED** [review]
